File: clients/ts/src/cross-language-worker.ts
import { createInterface } from "node:readline";

import { NetworkMutexClient, type LockHandle } from "./client.ts";

const host = process.env["LIVE_MUTEX_HOST"] ?? "127.0.0.1";
const port = Number(process.env["LIVE_MUTEX_PORT"] ?? "6970");
const lang = process.env["LMX_WORKER_LANG"] ?? "typescript";
const worker = process.env["LMX_WORKER_ID"] ?? `${lang}-0`;
const seed = BigInt(process.env["LMX_WORKER_SEED"] ?? "1");
const ops = Number(process.env["LMX_WORKER_OPS"] ?? "50");
const keyPrefix = process.env["LMX_FUZZ_KEY_PREFIX"] ?? "cross";
const keyCount = Number(process.env["LMX_FUZZ_KEY_COUNT"] ?? "5");

const TTL_MS = 60_000;

function makeRng(seed: bigint) {
  let x = (seed ^ 0x9e37_79b9_7f4a_7c15n) || 1n;
  return {
    below(n: number): number {
      x ^= x >> 12n;
      x ^= x << 25n;
      x ^= x >> 27n;
      x = BigInt.asUintN(64, x);
      return Number((x * 0x2545_f491_4f6c_dd1dn) % BigInt(n));
    },
  };
}

const stdinLines = createInterface({ input: process.stdin });
const ackQueue: string[] = [];
const waiters: Array<(line: string) => void> = [];
stdinLines.on("line", (line) => {
  const waiter = waiters.shift();
  if (waiter) waiter(line);
  else ackQueue.push(line);
});

async function waitAck(): Promise<void> {
  const line = ackQueue.shift() ?? await new Promise<string>((resolve) => waiters.push(resolve));
  if (line.trim() !== "ack") {
    throw new Error(`expected ack from harness, got ${JSON.stringify(line)}`);
  }
}

async function emit(event: unknown): Promise<void> {
  process.stdout.write(`${JSON.stringify(event)}\n`);
  await waitAck();
}

function chooseKeys(rng: ReturnType<typeof makeRng>, keys: string[]): string[] {
  const want = 2 + rng.below(Math.min(3, keys.length - 1));
  const pool = [...keys];
  const out: string[] = [];
  for (let i = 0; i < want && pool.length; i++) {
    out.push(pool.splice(rng.below(pool.length), 1)[0]!);
  }
  return [...new Set(out)].sort();
}

function tokensFor(handle: LockHandle): Record<string, number> {
  if (handle.kind === "single") return { [handle.key]: handle.fencingToken };
  return handle.fencingTokens;
}

async function holdBriefly(rng: ReturnType<typeof makeRng>): Promise<void> {
  const ms = rng.below(4);
  if (ms === 0) await Promise.resolve();
  else await new Promise((resolve) => setTimeout(resolve, ms));
}

async function grantReleaseExclusive(
  client: NetworkMutexClient,
  handle: LockHandle,
  rng: ReturnType<typeof makeRng>,
): Promise<void> {
  await emit({
    event: "grant",
    lang,
    worker,
    lockUuid: handle.lockUuid,
    kind: "exclusive",
    keys: handle.kind === "single" ? [handle.key] : handle.keys,
    tokens: tokensFor(handle),
  });
  await holdBriefly(rng);
  await emit({ event: "release", lang, worker, lockUuid: handle.lockUuid });
  await client.release(handle);
}

async function main(): Promise<void> {
  const rng = makeRng(seed);
  const keys = Array.from({ length: keyCount }, (_, i) => `${keyPrefix}-${i}`);
  const client = new NetworkMutexClient({ host, port, connectTimeoutMs: 5_000 });
  await client.connect();

  try {
    for (let op = 0; op < ops; op++) {
      const roll = rng.below(100);
      if (roll < 30) {
        const key = keys[rng.below(keys.length)]!;
        const h = await client.tryAcquire(key, { ttlMs: TTL_MS });
        if (h) await grantReleaseExclusive(client, h, rng);
      } else if (roll < 50) {
        const key = keys[rng.below(keys.length)]!;
        const h = await client.acquire(key, { ttlMs: TTL_MS, waitMs: TTL_MS });
        await grantReleaseExclusive(client, h, rng);
      } else if (roll < 65) {
        const h = await client.tryAcquireMany(chooseKeys(rng, keys), { ttlMs: TTL_MS });
        if (h) await grantReleaseExclusive(client, h, rng);
      } else if (roll < 75) {
        const h = await client.acquireMany(chooseKeys(rng, keys), { ttlMs: TTL_MS, waitMs: TTL_MS });
        await grantReleaseExclusive(client, h, rng);
      } else if (roll < 92) {
        const key = keys[rng.below(keys.length)]!;
        const h = await client.acquireRead(key);
        await emit({
          event: "grant",
          lang,
          worker,
          lockUuid: h.lockUuid,
          kind: "read",
          keys: [key],
          tokens: { [key]: h.fencingToken },
        });
        await holdBriefly(rng);
        await emit({ event: "release", lang, worker, lockUuid: h.lockUuid });
        await client.releaseRead(key);
      } else {
        const key = keys[rng.below(keys.length)]!;
        const h = await client.acquireWrite(key);
        await emit({
          event: "grant",
          lang,
          worker,
          lockUuid: h.lockUuid,
          kind: "write",
          keys: [key],
          tokens: { [key]: h.fencingToken },
        });
        await holdBriefly(rng);
        await emit({ event: "release", lang, worker, lockUuid: h.lockUuid });
        await client.releaseWrite(key);
      }
    }
  } finally {
    await client.close();
    stdinLines.close();
  }

  process.stdout.write(`${JSON.stringify({ event: "done", lang, worker, ops })}\n`);
}

main().catch((err) => {
  console.error(`[${worker}] ${err instanceof Error ? err.stack ?? err.message : String(err)}`);
  process.exit(1);
});

File: clients/dart/bin/cross_language_worker.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:dd_rust_network_mutex_client/client.dart';

final host = Platform.environment['LIVE_MUTEX_HOST'] ?? '127.0.0.1';
final port = int.parse(Platform.environment['LIVE_MUTEX_PORT'] ?? '6970');
final lang = Platform.environment['LMX_WORKER_LANG'] ?? 'dart';
final worker = Platform.environment['LMX_WORKER_ID'] ?? '$lang-0';
final seed = BigInt.parse(Platform.environment['LMX_WORKER_SEED'] ?? '1');
final ops = int.parse(Platform.environment['LMX_WORKER_OPS'] ?? '50');
final keyPrefix = Platform.environment['LMX_FUZZ_KEY_PREFIX'] ?? 'cross';
final keyCount = int.parse(Platform.environment['LMX_FUZZ_KEY_COUNT'] ?? '5');

const ttl = Duration(milliseconds: 60000);

class Rng {
  Rng(BigInt seed) : _x = seed ^ BigInt.parse('9e3779b97f4a7c15', radix: 16) {
    if (_x == BigInt.zero) _x = BigInt.one;
  }

  BigInt _x;

  int below(int n) {
    _x ^= _x >> 12;
    _x ^= _x << 25;
    _x ^= _x >> 27;
    _x &= (BigInt.one << 64) - BigInt.one;
    return ((_x * BigInt.parse('2545f4914f6cdd1d', radix: 16)) % BigInt.from(n))
        .toInt();
  }
}

final _ackQueue = <String>[];
final _ackWaiters = <Completer<String>>[];

StreamSubscription<String> _startAckReader() {
  return stdin.transform(utf8.decoder).transform(const LineSplitter()).listen((
    line,
  ) {
    if (_ackWaiters.isNotEmpty) {
      _ackWaiters.removeAt(0).complete(line);
    } else {
      _ackQueue.add(line);
    }
  });
}

Future<void> _waitAck() async {
  final line = _ackQueue.isNotEmpty
      ? _ackQueue.removeAt(0)
      : await (() {
          final c = Completer<String>();
          _ackWaiters.add(c);
          return c.future;
        })();
  if (line.trim() != 'ack') {
    throw StateError('expected ack from harness, got ${jsonEncode(line)}');
  }
}

Future<void> emit(Map<String, Object?> event) async {
  stdout.writeln(jsonEncode(event));
  await stdout.flush();
  await _waitAck();
}

List<String> chooseKeys(Rng rng, List<String> keys) {
  final want = 2 + rng.below(keys.length > 3 ? 3 : keys.length - 1);
  final pool = [...keys];
  final out = <String>[];
  for (var i = 0; i < want && pool.isNotEmpty; i++) {
    out.add(pool.removeAt(rng.below(pool.length)));
  }
  out.sort();
  return out.toSet().toList();
}

Future<void> holdBriefly(Rng rng) async {
  final ms = rng.below(4);
  if (ms > 0) await Future<void>.delayed(Duration(milliseconds: ms));
}

Future<void> grantReleaseExclusive(
  NetworkMutexClient client,
  Object handle,
  Rng rng,
) async {
  late final String lockUuid;
  late final List<String> keys;
  late final Map<String, int> tokens;
  if (handle is SingleLockHandle) {
    lockUuid = handle.lockUuid;
    keys = [handle.key];
    tokens = {handle.key: handle.fencingToken};
  } else if (handle is CompositeLockHandle) {
    lockUuid = handle.lockUuid;
    keys = handle.keys;
    tokens = handle.fencingTokens;
  } else {
    throw ArgumentError('unknown handle ${handle.runtimeType}');
  }
  await emit({
    'event': 'grant',
    'lang': lang,
    'worker': worker,
    'lockUuid': lockUuid,
    'kind': 'exclusive',
    'keys': keys,
    'tokens': tokens,
  });
  await holdBriefly(rng);
  await emit({
    'event': 'release',
    'lang': lang,
    'worker': worker,
    'lockUuid': lockUuid,
  });
  await client.release(handle);
}

Future<void> main() async {
  final ackReader = _startAckReader();
  final rng = Rng(seed);
  final keys = List.generate(keyCount, (i) => '$keyPrefix-$i');
  final client = await NetworkMutexClient.connect(host: host, port: port);
  try {
    for (var op = 0; op < ops; op++) {
      final roll = rng.below(100);
      if (roll < 30) {
        final h = await client.tryAcquire(
          keys[rng.below(keys.length)],
          ttl: ttl,
        );
        if (h != null) await grantReleaseExclusive(client, h, rng);
      } else if (roll < 50) {
        final h = await client.acquire(keys[rng.below(keys.length)], ttl: ttl);
        await grantReleaseExclusive(client, h, rng);
      } else if (roll < 65) {
        final h = await client.tryAcquireMany(chooseKeys(rng, keys), ttl: ttl);
        if (h != null) await grantReleaseExclusive(client, h, rng);
      } else if (roll < 75) {
        final h = await client.acquireMany(chooseKeys(rng, keys), ttl: ttl);
        await grantReleaseExclusive(client, h, rng);
      } else if (roll < 92) {
        final key = keys[rng.below(keys.length)];
        final h = await client.acquireRead(key);
        await emit({
          'event': 'grant',
          'lang': lang,
          'worker': worker,
          'lockUuid': h.lockUuid,
          'kind': 'read',
          'keys': [key],
          'tokens': {key: h.fencingToken},
        });
        await holdBriefly(rng);
        await emit({
          'event': 'release',
          'lang': lang,
          'worker': worker,
          'lockUuid': h.lockUuid,
        });
        await client.releaseRead(key);
      } else {
        final key = keys[rng.below(keys.length)];
        final h = await client.acquireWrite(key);
        await emit({
          'event': 'grant',
          'lang': lang,
          'worker': worker,
          'lockUuid': h.lockUuid,
          'kind': 'write',
          'keys': [key],
          'tokens': {key: h.fencingToken},
        });
        await holdBriefly(rng);
        await emit({
          'event': 'release',
          'lang': lang,
          'worker': worker,
          'lockUuid': h.lockUuid,
        });
        await client.releaseWrite(key);
      }
    }
  } finally {
    await client.close();
    await ackReader.cancel();
  }
  stdout.writeln(
    jsonEncode({'event': 'done', 'lang': lang, 'worker': worker, 'ops': ops}),
  );
  await stdout.flush();
}
